linera-views/DESIGN.md

```diff
@@ -0,0 +1,140 @@
+`base_key` as prefix.
+This leads us to introduce a trait `Context` that can be implemented with a specific
+`KeyValueStoreClient` client and a `base_key`.
+
+Another issue to consider is that we do not want to just store the values, but we
+also need to accommodate other features:
+* Administrative data (like journal) need to store some data starting from the `base_key`.
+* Some of the values need to be split into several smaller values.
+
+The rule for constructing are the following:
+* For the construction of `struct` objects of associated base key `base_key` we do the
+following: If the corresponding type has entries `entry_0`, ..., `entry_k` then the
+base key of the object associated with the `k`-th entry is `[base_key * * * *]` with `[* * * *]`
+the serialization of the number `k` considered as a `u32`.
+* For each view type, we have some associated data. Some are the hash, others the counts
+(e.g. for `LogView` and `QueueView`) and some the index. The corresponding enum is
+named `KeyTag` for each of those containers. For the index, the values are serialized.
+* For a given base key `base_key` the keys of the form `[base_key tag]` with tags in
+`0..MIN_VIEW_TAG` are reserved for administrative data.
+
+These constructions are done in order to satisfy the following property: The set of keys
+of a view is prefix-free. That is if a key is of the form `key = [a b]` then `a` is
+not a key.
+
+
+# Chosen implementation
+
+## API of the `KeyValueStoreClient`.
+
+The design of the key-value store client is done in the following way:
+* We can put a `(key, value)` and delete a `key`.
+* We can delete keys having a specific prefix.
+* We can retrieve the list of keys having a specific prefix. Similarly, we can
+retrieve the list of `(key, value)` whose keys have a specific prefix.
+* Also, the list of the keys are provided in the lexicographic order.
+
+## The journaling.
+
+We want to make sure that all operations are done on the database in an atomic
+way. That is the operations have to be processed completely. The idea is to
+store the journal in the `[base_key 0 * * * *]` keys with `[* * * *]` the
+serialization of an `u32` entry.
+Those keys are not colliding with the other keys since their value is of the
+form `[base_key tag *]` with `tag >= MIN_VIEW_TAG` and currently, we have
+`MIN_VIEW_TAG = 1`.
+
+The journaling allows to have operations in an atomic way. It is done in the
+following way:
+* When a `Batch` of operations needs to be processed, the operations are written
+into blocks that can be processed atomically. All together the blocks for the
+"journal". When the journal has been written then the counter is written and
+only at that point is the journal considered written.
+* Each block is then processed atomically. If an interruption occurs then
+the block entries would be processed after accessing the database the next time.
+
+# The `ValueSplittingKeyValueStoreClient` type.
```

Comment:
`## Splitting large values across keys` (to be consistent with `## Journaling`)
Otherwise change the previous title to be consistent with this one

Reply:
Ok, done.